### PR TITLE
`WidgetState` mapping

### LINKED ISSUE #143488: Customize floatingLabelStyle for error in InputDecorationTheme
### Use case

I want to customize my TextField entirely through themeData so I don't have to create a specific widget as a wrapper for it.

Right now my form field looks like this: 
image
image
image
image

InputDecorationTheme doesn't expose any parameters to customize textStyle of floatingLabel when validating textField with an error. I can't change the color of the floatingLabel when focused on the textField, and unfocusing it makes the label take the default dark theme error color.

image

### Proposal

Give inputDecorationTheme optional parameters focusedErrorFloatingLabelStyle and errorFloatingLabelStyle.

### LINKED ISSUE #146042: Feature Request: a more readable `WidgetStateProperty` constructor that maintains flexibility
As explained in a [Decoding Flutter video](https://youtu.be/CylXr3AF3uU?si=hFJur9HYpYXXNyIp&t=43), the `WidgetStateProperty` class (which until recently was called MaterialStateProperty) allows for styling based on multiple different factors without an enormous parameter list.

<br>

## Problem

Let's look at some common use cases:
- Check `WidgetState`s one at a time
- Check multiple `WidgetState`s to see if any of them match
- Check multiple `WidgetState`s to see if all of them match

<details open> <summary>code snippet (click to collapse)</summary>

```dart
// one at a time
WidgetStateProperty.resolveWith<Color>((states) {
  if (states.contains(WidgetState.disabled)) {
    return Colors.grey;
  } else if (states.contains(WidgetState.error)) {
    return Colors.red;
  } else if (states.contains(WidgetState.pressed)) {
    return Colors.blueAccent;
  } else if (states.contains(WidgetState.focused)) {
    return Colors.blue;
  } else if (states.contains(WidgetState.hovered)) {
    return Colors.blueGrey;
  }
  return Colors.black;
});

const activeStates = [
  WidgetState.pressed,
  WidgetState.hovered,
  WidgetState.focused,
];

// see if any match
WidgetStateProperty.resolveWith<Color>((states) {
  if (activeStates.any(states.contains)) {
    if (states.contains(WidgetState.error) {
      return Colors.red;
    }
    return Colors.blue;
  }
  return Colors.black;
});

// see if all match
WidgetStateProperty.resolveWith<Color>((states) {
  if (states.containsAll(activeStates)) {
    if (states.contains(WidgetState.error) {
      return Colors.red;
    }
    return Colors.blue;
  }
  return Colors.black;
});
```

</details>

#### Pros
- effective

#### Cons
- ugly

With 8 different enum values, we have 2⁸ different `Set<WidgetState>` possibilities, so finding a `WidgetPropertyResolver` implementation that isn't messy or limited in scope is challenging.

<br>

### `Set` pattern matching

<details> <summary>code snippet</summary>

```dart
// one at a time
WidgetStateProperty.resolveWith<Color>((states) => switch (states) {
  {WidgetState.disabled} => Colors.grey,
  {WidgetState.error}    => Colors.red,
  {WidgetState.pressed}  => Colors.blueAccent,
  {WidgetState.focused}  => Colors.blue,
  {WidgetState.hovered}  => Colors.blueGrey,
  _                      => Colors.black,
});

// see if any match
WidgetStateProperty.resolveWith<Color>((states) => switch (states) {
  {WidgetState.pressed || WidgetState.hovered || WidgetState.focused, WidgetState.error} => Colors.red,
  {WidgetState.pressed || WidgetState.hovered || WidgetState.focused} => Colors.blue,
  _ => Colors.black,
});

// see if all match
WidgetStateProperty.resolveWith<Color>((states) => switch (states) {
  {WidgetState.pressed, WidgetState.hovered, WidgetState.focused, WidgetState.error} => Colors.red,
  {WidgetState.pressed, WidgetState.hovered, WidgetState.focused} => Colors.blue,
  _ => Colors.black,
});
```

</details>

#### Pros
- super clean & readable 😍

#### Cons
- This isn't supported in the Dart language 😢<br>There's an [open issue](https://github.com/dart-lang/language/issues/3296) for it though; maybe we could just keep our fingers crossed.
<br>

### Convert to `Map`

```dart
T resolveWithMap<T>(T function(Map<WidgetState, bool> resolve)) => resolveWith((states) {
  final stateMap = {for (final state in WidgetStates) state: states.contains(state)};
  return resolve(stateMap);
});
```

<details> <summary>code snippet</summary>

```dart
// one at a time
WidgetStateProperty.resolveWithMap<Color>((stateMap) => switch (stateMap) {
  {WidgetState.disabled: true} => Colors.grey,
  {WidgetState.error:    true} => Colors.red,
  {WidgetState.pressed:  true} => Colors.blueAccent,
  {WidgetState.focused:  true} => Colors.blue,
  {WidgetState.hovered:  true} => Colors.blueGrey,
  _                            => Colors.black,
});

// see if any match
WidgetStateProperty.resolveWithMap<Color>((stateMap) {
  switch (stateMap) {
    case {WidgetState.pressed: true}:
    case {WidgetState.hovered: true}:
    case {WidgetState.focused: true}:
      if (states.contains(WidgetState.error) {
        return Colors.red;
      }
      return Colors.blue;
    default:
      return Colors.black;
  }
});

// see if all match
WidgetStateProperty.resolveWithMap<Color>((stateMap) {
  switch (stateMap) {
    case {WidgetState.pressed: false}:
    case {WidgetState.hovered: false}:
    case {WidgetState.focused: false}:
      return Colors.black;
    case {WidgetState.error: true}:
      return Colors.red;
    default:
      return Colors.blue;
  }
});
```

</details>

#### Pros
- Works with the current pattern matching features
- The ability to check for either `true` or `false` helps avoid repetition

#### Cons
- A tinge of sadness—in most situations, it's not quite as concise as `Set` pattern matching would be

<br>

### Enhanced `enum`

Instead of just trying to improve the way we parse the set, we could also try to improve `WidgetState` directly.

```dart
enum WidgetState with SpicyMixin { ... }
```

<details> <summary>code snippet</summary>

```dart
// one at a time
WidgetStateProperty.map<Color>({
  WidgetState.disabled: Colors.grey,
  WidgetState.error:    Colors.red,
  WidgetState.pressed:  Colors.blueAccent,
  WidgetState.focused:  Colors.blue,
  WidgetState.hovered:  Colors.blueGrey,
  WidgetState.any:      Colors.black, // "any" is a static const member, not an enum value
});

final anyActive = WidgetState.pressed | WidgetState.hovered | WidgetState.focused;

// see if any match
WidgetStateProperty.map<Color>({
  anyActive & WidgetState.error: Colors.red,
  anyActive: Colors.blue,
  ~anyActive: Colors.black,
});

final allActive = WidgetState.pressed & WidgetState.hovered & WidgetState.focused;

// see if all match
WidgetStateProperty.map<Color>({
  allActive & WidgetState.error: Colors.red,
  allActive: Colors.blue,
  ~allActive: Colors.black,
});
```

</details>

#### Pros
- Even more concise and readable than `Set` pattern matching
- Even more flexible than `Map` pattern matching
- Syntax is similar to named arguments

#### Cons
- The syntax still feels kinda wonky, though.

### PR DESCRIPTION
**Pull request by nate-thegrate**
This pull request implements [enhanced enum](https://dart.dev/language/enums#declaring-enhanced-enums) features for the new `WidgetState` enum, in order to improve the developer experience when creating and using `WidgetStateProperty` objects.

`WidgetState` now has a `.matchesSet()` method:

```dart
// identical to "states.contains(WidgetState.error)"
final bool hasError = WidgetState.error.isSatisfiedBy(states);
```

This addition allows for wide variety of `WidgetStateProperty` objects to be constructed in a simple manner.

<br><br>

```dart
// before
final style = MaterialStateTextStyle.resolveWith((states) {
  if (states.contains(MaterialState.error)) {
    return const TextStyle(color: Colors.red);
  } else if (states.contains(MaterialState.focused)) {
    return const TextStyle(color: Colors.blue);
  }
  return const TextStyle(color: Colors.black);
});


// after
const style = WidgetStateTextStyle.fromMap({
  WidgetState.error:   TextStyle(color: Colors.red),
  WidgetState.focused: TextStyle(color: Colors.blue),
  WidgetState.any:     TextStyle(color: Colors.black),
});
```

```dart
// before
final color = MaterialStateProperty.resolveWith((states) {
  if (states.contains(MaterialState.focused)) {
    return Colors.blue;
  } else if (!states.contains(MaterialState.disabled)) {
    return Colors.black;
  }
  return null;
});


// after
final color = WidgetStateProperty<Color?>.fromMap({
  WidgetState.focused:   Colors.blue,
  ~WidgetState.disabled: Colors.black,
});
```

```dart
// before
const activeStates = [MaterialState.selected, MaterialState.focused, MaterialState.scrolledUnder];

final color = MaterialStateColor.resolveWith((states) {
  if (activeStates.any(states.contains)) {
    if (states.contains(MaterialState.hovered) {
      return Colors.blueAccent;
    }
    return Colors.blue;
  }
  return Colors.black;
});


// after
final active = WidgetState.selected | WidgetState.focused | WidgetState.scrolledUnder;

final color = WidgetStateColor.fromMap({
  active & WidgetState.hovered: Colors.blueAccent,
  active:  Colors.blue,
  ~active: Colors.black,
});
```

<br>

(fixes #146042, and also fixes #143488)

### CODE REVIEW

**Comment by cedvdb** [comment]
imo `WidgetState.isError(states)` would be more intuitive than `WidgetState.error(states)`

**Comment by nate-thegrate** [comment]
@cedvdb this is a really good point.

In order to make the mapping work, I [added a `call()` method](https://github.com/nate-thegrate/flutter/blob/widget-state-mapping/packages/flutter/lib/src/widgets/widget_state.dart#L156) to `WidgetState`, and it has the nice side effect of providing a syntax that's more concise than `states.contains()`. Since it's a `call()` method for the enum value rather than a static class method, this would be a change to the enum name:

```dart
enum WidgetState { isHovered, isFocused, isPressed, ... isError }
```

<br>

Luckily, there's already a way to add an intuitive `states.isError` syntax, as mentioned in a [Flutter YouTube video](https://youtu.be/CylXr3AF3uU?t=449). And with this PR, there's no need to directly reference `states` at all:

```dart
WidgetStateProperty.map({WidgetState.error: Colors.red, WidgetState.any: Colors.black});
```

**Review by MitchellGoodwin — COMMENTED** [review]
Thank you for this PR! I think this is a great solution, that covers simple cases and complicated combinations well. I have some questions I'd like to talk about, but no complaints otherwise.

**Review by MitchellGoodwin — APPROVED** [review]
LGTM with one follow up on naming. Thanks again for this PR, it's a great addition.

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 1]
```diff
@@ -171,8 +244,8 @@ abstract class WidgetStateColor extends Color implements WidgetStateProperty<Col
   static const WidgetStateColor transparent = _WidgetStateColorTransparent();
 }
 
-class _WidgetStateColor extends WidgetStateColor {
-  _WidgetStateColor(this._resolve) : super(_resolve(_defaultStates).value);
+class _WidgetStateColorWith extends WidgetStateColor {
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/widget_state.dart:247]: I'd never thought about how the `with` keyword could make this confusing. Sounds great!

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 2]
```diff
@@ -9,6 +9,64 @@ import 'package:flutter/services.dart';
...
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const WidgetStateMapKey();
+
+  /// A function used by [WidgetStateProperty.map] to check whether
+  /// each key matches the current set of states.
+  bool call(Set<WidgetState> states);
+}
+
+/// A private class, used in [WidgetStateOperators].
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/widget_state.dart:26]: trivial nit: this should be `//` not `///` per the style guide, because we wouldn't want to publish the documentation in its current state if we made the class public. see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-for-public-quality-private-documentation

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 3]
```diff
@@ -9,6 +9,64 @@ import 'package:flutter/services.dart';
...
+    return _WidgetStateOperation((Set<WidgetState> states) => _this(states) || other(states));
+  }
+
+  /// Takes a [WidgetStateMapKey] and applies the logical "not".
+  WidgetStateMapKey operator ~() {
+    return _WidgetStateOperation((Set<WidgetState> states) => !_this(states));
+  }
+}
+
+/// A private class, used to create [WidgetState.any].
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/widget_state.dart:62]: same here

**Comment by MitchellGoodwin** [comment]
The Google testing looks related. In one spot, there is a set of only a few WidgetStates.
Pseudo code it looks like this.
```
const certainStates = <WidgetState>{
   WidgetState.pressed,
   Widget.State.hovered,
}
```

It threw this error over this:
Error: Constant evaluation error: const certainStates = <WidgetState>{

Context: The element 'WidgetState {}' conflicts with another existing element in the set.
 - 'WidgetState' is from 'package:flutter/src/widgets/widget_state.dart'
 WidgetState.focused,
 
 Context: While analyzing: const interactiveStates = <WidgetState>{

**Comment by MitchellGoodwin** [comment]
It's possible there are other issues, but at least most of them seem to be around that set

**Comment by nate-thegrate** [comment]
Dang. Guess I'll stop trying to repeatedly run the checks 🙂

**Comment by MitchellGoodwin** [comment]
This PR when merged (#148804) is going to add a similar `const` set of `WidgetState`s that should throw the same error as what's in Google testing. We would need to create a way to allow having a constant set of states, because that would be a bit of a bummer of a regression.

**Comment by nate-thegrate** [comment]
> We would need to create a way to allow having a constant set of states, because that would be a bit of a bummer of a regression.

@MitchellGoodwin I have some good news: `const Set<WidgetState>` still works!

Just to double-check, I checked out my branch and did a `flutter run` with the following:

```dart
// main.dart

void main() {
  runApp(const MaterialApp(home: Scaffold()));

  const WidgetStateMap<String> widgetStateMap = {
    WidgetState.selected: 'it works!',
  };
  const Set<WidgetState> states = {
    WidgetState.selected,
    WidgetState.hovered,
  };

  final doesItWork = WidgetStateProperty.map(widgetStateMap).resolve(states);
  print(doesItWork); // it works!
}
```

I'm looking forward to #148804 being merged; at that point hopefully we can copy those changes into this branch and see what's going on.

**Comment by nate-thegrate** [comment]
Super random, but I wanted to double-check that `WidgetStateProperty.map` was a `const` constructor (since `const` WidgetStateMaps will probably be a common occurrence), and it turns out that it wasn't!

<br>

As far as Google Testing errors, we could continue waiting for #148804, but this has been on my mind for a while and I think I might have an idea.

If a `const` set of states were causing the issues, then the button tests would also fail:

https://github.com/flutter/flutter/blob/c63339866d4f7477777e7377bf044f14b743f71c/packages/flutter/test/material/filled_button_test.dart#L2256-L2258

My best guess is that it's an import error: now that the enum has been changed on my branch, maybe the test file is trying to import 2 different versions of the same type at once. (I ran into this error before when I made a class that had a name overlap with the Material library, and then tried to import both `material.dart` and the file I made.)

A super quick & dirty fix would be to give each import a different prefix to guarantee that there won't be any overlaps.

(Also, I feel like I should apologize for this attempt to boss you around. I really wish I could just see the test output directly and try to fix it myself, but I guess it is what it is.)

**Comment by MitchellGoodwin** [comment]
> My best guess is that it's an import error: now that the enum has been changed on my branch, maybe the test file is trying to import 2 different versions of the same type at once.

Took a look through. It's not a unit test that's failing but a build error on a smoketest. The set that's being a problem is called `interactiveStates`, which we do use in some tests and examples, but they are scopped in a way so I do not think they would conflict.

Looking at the stack trace, that set is first called for this kind of function:
```final numberOfCurrentInteractions =
states.intersection(interactiveStates).length;
```
One file imports Material, the other one, the Widgets library. Neither import much more, both files are high level.

**Comment by nate-thegrate** [comment]
Dang, I guess I'll try digging around on the internet for answers.

I really appreciate you looking into it.

**Comment by nate-thegrate** [comment]
I think I was able to replicate the error message:

```
lib/test.dart:3:27: Error: Constant evaluation error:
const interactiveStates = {
                          ^
lib/test.dart:5:15: Context: The element 'WidgetState {}' conflicts with another existing element in the set.
 - 'WidgetState' is from 'package:flutter/src/widgets/widget_state.dart' ('/C:/…/widget_state.dart').
  WidgetState.focused,
              ^
lib/test.dart:3:7: Context: While analyzing:
const interactiveStates = {
      ^
```

```dart
// test.dart

import 'package:flutter/material.dart';

const interactiveStates = {
  WidgetState.focused,
  WidgetState.focused,
};
```

I have no clue how it relates to this PR.

**Comment by MitchellGoodwin** [comment]
@nate-thegrate Google testing is failing on different tests then before, but they look odd. Can you rebase so we can see if it's not related to something else? One of the tests is a golden file test with a difference under 0.1% so I can't see what's different, and the other looks more like a difference in backend responses that I wouldn't think is related to WidgetState. If it fails again I'll dig deeper.

**Comment by MitchellGoodwin** [comment]
Looks like it's the same errors as before, with the constant evaluation error.

**Comment by goderbauer** [comment]
This is a very clever implementation! However, I am concerned that it makes `WidgetState` and the code it is used in harder to comprehend. If you see code like `states.contains(MaterialState.error)` it's immediately clear what's happening. That's not true for `WidgetState.error(states)`. It's not obvious at all that the latter returns a boolean, and if I want to figure out what this does it gets even more confusing: At first glance, I would have expected that `WidgetState.error` is some kind of a static function, but if I click through I confusingly end up at an enum value - and I am sort of stuck there. There is no indication what "invoking" an enum value would get me. I have to remember the dart magic trick that implementing a `call` method makes any object callable and then look at the implementation of said `call` method to figure out what's actually going on in this code snippet. In the original (`states.contains(MaterialState.error)`) it's clear right there what's happening. To understand the new world (`WidgetState.error(states)`) I have to do two hops (one of which requires specialized knowledge about dart's magic - and rarely used - `call` method) to get an understanding of what the code does.

This is also true for the proposed `map` method: It's very compact and clever! But it's also way harder to understand what's happening there.

tl;dr: I admire the technical skills and clever implementation behind this. It's genius! Ultimately, I think we shouldn't do this, though, as it will hurt readability and understandability of the code.

**Comment by nate-thegrate** [comment]
@goderbauer I really appreciate this feedback!

I received similar feedback for #148449, and I was happy to close it (since I was feeling uneasy about it as well).
Likewise, after reading your thoughts about this PR, I agree that the implementation is severely lacking in readability and understandability.

However: rather than close this right away, I'd like to take some time this evening to work through the issues you mentioned.
I fervently believe that an improved version of this PR would be immensely useful for the people who want a more straightforward `WidgetPropertyResolver` implementation.

**Comment by nate-thegrate** [comment]
When I look at enum values in VS Code, they have consistent syntax highlighting, which creates a clear visual distinction between a "call" method and a static class method.

Unfortunately, other places (such as GitHub) don't behave the same way. And even if they did, I admit it was still was a bad idea to use `call()` here, due to its impact on readability.

<br>

I've updated the enum to use a named method `matchesSet`, to prevent the appearance of directly invoking the object.
I think it makes the new constructor's `resolve` method significantly easier to follow:

```dart
for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
  if (entry.key.matchesSet(states)) {
    return entry.value;
  }
}
```

<br>

I also gave the [`matchesSet` doc comments](https://github.com/flutter/flutter/blob/0afd932a69e10fce59150a393ecdbe856a3eb20f/packages/flutter/lib/src/widgets/widget_state.dart#L22-L39) a clearer explanation:

> Whether this key should apply, given the current [states].
> 
> If the key is a single [WidgetState] object, it matches the set if the set contains the object.
> 
> The key could also be created using one or more operators, for example:
> ```dart
> final WidgetStateMapKey mapKey = WidgetState.focused | WidgetState.hovered;
> ```
> 
> In the above case, the output of `mapKey.matchesSet(states)` is equivalent to:
> ```dart
> states.contains(WidgetState.focused) || states.contains(WidgetState.hovered);
> ```

<br>

I believe that the name change and added documentation enhances the readability and maintainability of the `WidgetStateMap` interface.

**Comment by nate-thegrate** [comment]
### ✨ 🙏 Let us sing Hosanna, for miracles cometh from on high 🙏 ✨

<br>

I do not understand it whatsoever, but here's what I've discovered:
- Adding an operator to `WidgetState` → works (but doesn't allow chaining operators)
- Adding operators to both `WidgetStateMapKey` and `WidgetState` → doesn't work
- Putting operators in a mixin, applied to both classes → doesn't work
- Making an `extension on WidgetStateMapKey` → works!

This would have been much more straightforward if enums could extend other classes, but I guess we gotta work with what we got.

I didn't want to throw on `autosubmit` right away: extensions are used in multiple places throughout the repo, but the [style guide](https://github.com/flutter/flutter/blob/fa70e610cc244750b07321c466385d821533ee68/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-extension) recommends against them.

It isn't possible to subclass an enum, so the only way that this extension would make someone upset is if they implemented a different subclass of `WidgetStateMapKey` and they didn't like the operator logic.
In light of the fact that Google Testing wonkiness makes an `extension` the only working option here, I believe this is a situation where it's warranted. (If someone could give a thumbs-up or something, that would be much appreciated.)

**Comment by goderbauer** [comment]
Awesome. Before we submit this, I'd like to take another look at this from a readability and comprehension point of view. It's in my queue and I hope to get to it either this or early next week.

**Comment by nate-thegrate** [comment]
> Before we submit this, I'd like to take another look at this from a readability and comprehension point of view.

Sounds great! I might do that as well :)

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 4]
```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
...
+
+// A private class, used to create the [WidgetStateProperty.map] constructor.
+class _WidgetStateMapper<T> implements WidgetStateProperty<T> {
+  const _WidgetStateMapper(this.map);
+
+  final WidgetStateMap<T> map;
+
+  @override
+  T resolve(Set<WidgetState> states) {
+    for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_state.dart:819]: I am missing documentation stating in which order it will try to match the entries of the provided map. Can you add this to some prominent places?

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 5]
```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
...
+      'Consider using "WidgetStateProperty<$T?>.map()", '
+      'or adding the "WidgetState.any" key to this map.',
+    );
+
+    try {
+      return null as T;
+    } on TypeError {
+      throw ArgumentError(errorMessage);
+    }
+  }
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_state.dart:839]: I wonder if we shouldn't allow implicitly returning null. Would it maybe be better to explicitly require an entry for null in the map if you want it to ever resolve to null?

Would be interesting to look at some use cases to see how common the null case is and if making it explicit would just add a ton of boilerplate.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/widget_state.dart:839]: Also, this implicit null behavior needs to be documented somewhere clearly.

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 6]
```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
...
 }
 
+/// A [Map] used to resolve to a single value of type `T` based on
+/// the current set of Widget states.
+///
+/// {@template flutter.widgets.WidgetStateProperty.WidgetStateMap}
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.map(<WidgetStateMapKey, Color?>{
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/widget_state.dart:759]: Alternatively: `WidgetStateProperty<Color?>.fromMap` to emphasize that this creates a  WidgetStateProperty from a map and doesn't map values to another value.

File: packages/flutter/lib/src/widgets/widget_state.dart (hunk 4 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/widget_state.dart:851]: Good point. Since this is a private class's method, I'll document the behavior in each constructor that uses it.

File: packages/flutter/lib/src/widgets/widget_state.dart (hunk 5 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/widget_state.dart:868]: > this implicit null behavior needs to be documented somewhere clearly.

Absolutely!

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 7]
```diff
@@ -8,6 +8,89 @@ import 'package:flutter/services.dart';
...
+
+/// This class allows [WidgetState] enum values to be combined
+/// using [WidgetStateOperators].
+///
+/// A [Map] with [WidgetStateMapKey] objects as keys can be used
+/// in the [WidgetStateProperty.fromMap] constructor to resolve to
+/// one of its values, based on the first key that [isSatisfiedBy]
+/// the current set of states.
+///
+/// {@macro flutter.widgets.WidgetStateProperty.WidgetStateMap}
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/widget_state.dart:21]: As far as the order of examples in this doc macro, my thoughts are that you don't _need_ to understand the operators in order to make a `WidgetStateMap`, but you definitely would need to understand the basics of a `WidgetStateMap` in order to know how to use the operators.

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 8]
```diff
@@ -8,6 +8,80 @@ import 'package:flutter/services.dart';
...
+abstract interface class WidgetStateMapKey {
+  /// Whether this key should apply, given the current [states].
+  ///
+  /// If the key is a single [WidgetState] object,
+  /// it matches the set if the set contains the object.
+  ///
+  /// The key could also be created using one or more operators, for example:
+  ///
+  /// ```dart
+  /// final WidgetStateMapKey mapKey = WidgetState.focused | WidgetState.hovered;
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/widget_state.dart:32]: After thinking it over for a while, I do really like the parallelism with `BoxConstraints`.

I'll go ahead and change the class name accordingly.

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 9]
```diff
@@ -8,6 +8,89 @@ import 'package:flutter/services.dart';
...
+/// using [WidgetStateOperators].
+///
+/// A [Map] with [WidgetStatesConstraint] objects as keys can be used
+/// in the [WidgetStateProperty.fromMap] constructor to resolve to
+/// one of its values, based on the first key that [isSatisfiedBy]
+/// the current set of states.
+///
+/// {@macro flutter.widgets.WidgetStateProperty.WidgetStateMap}
+abstract interface class WidgetStatesConstraint {
+  /// Whether the current [states] satisfy this object's criteria.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_state.dart:23]: nit:
```suggestion
  /// Whether the provided [states] satisfy this object's criteria.
```

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 10]
```diff
@@ -519,12 +686,24 @@ class _WidgetStateTextStyle extends WidgetStateTextStyle {
...
+  /// {@template flutter.widgets.WidgetStateProperty.fromMap}
+  /// This constructor's [resolve] method finds the first [MapEntry] whose
+  /// key is satisfied by the set of states, and returns its associated value.
+  /// {@endtemplate}
+  ///
+  /// Returns `null` if no keys match, or if [T] is non-nullable,
+  /// the method throws an [ArgumentError].
+  ///
+  /// {@macro flutter.widgets.WidgetState.any}
+  const factory WidgetStateProperty.fromMap(WidgetStateMap<T> map) = _WidgetStateMapper<T>;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_state.dart:706]: I suspect that this is a common entry point to WidgetState; maybe include the general example in this doc comment as well to give a general overview of how this all works.

File: packages/flutter/lib/src/widgets/widget_state.dart [hunk 11]
```diff
@@ -594,6 +780,95 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
...
 }
 
+/// A [Map] used to resolve to a single value of type `T` based on
+/// the current set of Widget states.
+///
+/// {@template flutter.widgets.WidgetStateProperty.WidgetStateMap}
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.fromMap(<WidgetStatesConstraint, Color?>{
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_state.dart:790]: Maybe leave a comment on here to explain that the map passed to fromMap is of type `WidgetStateMap<Color?>` for people wondering why this example is given on WidgetStateMap.

I agree with your comment above that assigning it to a variable is not a good idea.

**Review by goderbauer — APPROVED** [review]
LGTM after some minor doc adjustments (see comments).

**Comment by nate-thegrate** [comment]
Third time's the charm! 🤞
